### ISSUE TITLE
Training Float32 model and Half model simultaneously

### ISSUE THREAD

**Issue by npmhung** [NVIDIA/apex#277]
Hi,
As stated in the title, I want to train my network including model 1 and model 2.
For model 1, I can train it using the half precision. However, for model 2, because sometimes I have to perform sum operation over the feature maps, the fact that makes it easy to go overflow. Therefore, I want to keep training that under float32.
Does apex support this kind of training?

**Comment by ptrblck** [comment]
Hi @npmhung,
I think you could use the context manager `with amp.handle.disable_casts()` to keep the critical model in FP32.
Here is a small example using this approach:
```python
import torch
import torch.nn as nn
import torch.nn.functional as F

from apex import amp


class MyCriticalModule(nn.Module):
    def __init__(self):
        super(MyCriticalModule, self).__init__()
        
    def forward(self, x):
        print(x.type())
        x = x.view(x.size(0), -1).sum(1) + 65504. # add largest FP16 number to trigger overflow
        print(x.type())
        print(x)
        return x
        

class MyCriticalModel(nn.Module):
    def __init__(self):
        super(MyCriticalModel, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 3, 1, 1)
        self.pool1 = nn.MaxPool2d(2)
        self.conv2 = nn.Conv2d(6, 12, 3, 1, 1)
        self.pool2 = nn.MaxPool2d(2)
        self.crit = MyCriticalModule()
        self.fc1 = nn.Linear(1, 1)
        
    def forward(self, x):
        x = self.pool1(F.relu(self.conv1(x)))
        x = self.pool2(F.relu(self.conv2(x)))
        x = self.crit(x)
        x = self.fc1(x)
        
        return x

class MyModel(nn.Module):
    def __init__(self):
        super(MyModel, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 3, 1, 1)
        self.pool1 = nn.MaxPool2d(2)
        self.conv2 = nn.Conv2d(6, 12, 3, 1, 1)
        self.pool2 = nn.MaxPool2d(2)
        self.fc1 = nn.Linear(12*24*24, 1)
        
    def forward(self, x):
        x = self.pool1(F.relu(self.conv1(x)))
        x = self.pool2(F.relu(self.conv2(x)))
        x = x.view(x.size(0), -1)
        x = self.fc1(x)
        
        return x


device = 'cuda'
critical_model = MyCriticalModel().to(device)
model = MyModel().to(device)

critical_model, model = amp.initialize([critical_model, model], opt_level='O1')
x = torch.randn(1, 3, 96, 96, device=device)

with amp.handle.disable_casts():
    critical_output = critical_model(x)
output = model(x)

print(critical_output)
print(output)
```
Note that if you remove the context manager and just run `critical_output = critical_model(x)` you will most likely see an output of `torch.tensor(-inf)`, since I've triggered an overflow in the forward method.

Would this work for you or did I misunderstand your use case?

**Comment by mcarilli** [comment]
@npmhung Also, do you want to keep the entire model in fp32, or just the sum operation?  Because with opt_level="O1", Amp will carry out all sums in FP32 anyway.  It's possible that
```
[model1, model2], optimizer = amp.initialize([model1, model2, optimizer, opt_level="O1")
```
is all you need to do.

**Comment by npmhung** [comment]
@mcarilli that's exactly what I need. I'm struggling with the overflow when performing the sum over the spatial dim of feature map. Hence, I tried to convert to float and half manually, but sometimes it still resulted in overflow problem.
@ptrblck this is great. With this context manager, I can make my model more flexible.

**Comment by mcarilli** [comment]
@npmhung Note that this context manager only affects opt_level="O1".  Also, based on your what you've said so far, it's possible setting `opt_level="O1" is all you need to do, and you don't need to use the manual context manager at all.

Originally, I made a minor syntax error in https://github.com/NVIDIA/apex/issues/277#issuecomment-487624970.  I said 
```
[model1, model2], optimizer = amp.initialize([model1, model2, optimizer, opt_level="O1")
```
(forgot to close the braces on the list of models passed to `amp.initialize`).  Correct version is
```
[model1, model2], optimizer = amp.initialize([model1, model2], optimizer, opt_level="O1")
```
(also corrected in https://github.com/NVIDIA/apex/issues/277#issuecomment-487624970).

**Comment by John1231983** [comment]
@mcarilli : I want to train all network with float32 but I want to use apex to persevere gradient flow
```
 with amp.scale_loss(loss1, optimizer1) as scaled_loss:
```
Which setting should I used? AFAIK, the opt_level="01" will convert the model to F16

**Comment by mcarilli** [comment]
Gradient underflow and overflow are highly unlikely if the entire network is fp32.  If the entire network is fp32, you don't need apex at all.
